tower-http/src/classify_response.rs

```diff
@@ -0,0 +1,82 @@
+use http::{Response, StatusCode};
+use std::fmt;
+
+/// Classify requests results as either success or failure.
+///
+/// Not all services can rely entirely on the status code to determine if its a success or failure.
+/// gRPC for example has its own way of communicating that and always uses `200 OK`.
+pub trait ClassifyResponse<Body, Err> {
+    /// The type used to classify successful responses.
+    type OkClass;
+
+    /// The type used to classify failed responses.
+    type ErrClass;
+
+    /// Do the classification.
+    fn classify_request_result(
+        &self,
+        request_result: &Result<Response<Body>, Err>,
+    ) -> ResponseClassification<Self::OkClass, Self::ErrClass>;
+}
+
+/// Was a response a success or a failure?
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum ResponseClassification<T, E> {
+    /// The response was a success.
+    ///
+    /// `T` is a value produced by doing the classification. Could for example be the status code
+    /// that a middleware uses.
+    Ok(T),
+
+    /// The response was a failure.
+    ///
+    /// `E` is a value produced by doing the classification. Could for example be an error.
+    Err(E),
+}
+
+/// The default [`ClassifyResponse`] that some middlewares use.
+///
+/// It classifies `Ok(response)` as success unless the status code is `5xx`. Errors are always
+/// classified as errors.
+#[derive(Debug, Clone, Default)]
+pub struct DefaultHttpResponseClassifier {
+    _priv: (),
+}
+
+impl DefaultHttpResponseClassifier {
+    /// Create a new [`DefaultHttpResponseClassifier`].
+    pub fn new() -> Self {
+        Self { _priv: () }
+    }
+}
+
+impl<B, E> ClassifyResponse<B, E> for DefaultHttpResponseClassifier
+where
+    E: fmt::Display,
+{
+    type OkClass = StatusCode;
+    type ErrClass = DefaultErrorClassification;
+
+    fn classify_request_result(
+        &self,
+        request_result: &Result<Response<B>, E>,
+    ) -> ResponseClassification<StatusCode, DefaultErrorClassification> {
+        match request_result {
+            Ok(res) if res.status().is_server_error() => {
+                let err = format!("Server error. status={}", res.status());
```

Comment:
this will allocate a string every time a server error is returned. that feels like a _lot_ of overhead that I don't think is really necessary. what do you think about changing `DefaultErrorClassification` to something more like this:
```rust
pub struct DefaultErrorClassification<E> {
    inner: DefaultErrorKind<E>,
}

enum DefaultErrorKind<E> {
    Err(E),
    Status(http::Status),
}

impl<E: fmt::Display> fmt::Display for DefaultErrorClassification<E> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.inner {
            DefaultErrorKind::Status(s) => write!(f, "server error. status={}", s),
            DefaultErrorKind::Err(e) => write!(f, "{}", e),
        }
    }
}
```
or something like that. This way, we don't allocate strings at all, and only format the errors/status codes when needed. Also, you could easily add an `Error` impl for `DefaultErrorClassification` that allows downcasting to the original `Error`.

Reply:
Yeah I agree having to allocate a string sucks. However since the classify method receives `&Result<Response<_>, E>` we cannot get an owned `E` to put into the return value. That introduces a lifetime which spread everywhere. Not sure if there is an elegant solution to that.